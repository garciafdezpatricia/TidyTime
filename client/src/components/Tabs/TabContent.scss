@use "../../../styles/variables" as *;

.tab-content-container {
    grid-row: 2/-1;
    grid-column: 1/-1;
    height: 90%;
    max-height: 90%;
    padding-left: 1rem;
}

.tab-content-empty {
    grid-row: 2/-1;
    grid-column: 1/-1;
    height: 90%;
    max-height: 90%;
    padding-left: 1rem;
    display: flex;
    align-items: center;
    justify-content: center;
    background-image: url("./../../../public/add-task-list.svg");
    background-size: contain;
    background-repeat: no-repeat;
    background-position: center;
}

.content {
    display: none;
}

.active-content {
    height: 90%;
    width: 100%;
    overflow-y: auto;
    overflow-x: hidden;
    max-height: 25rem;
}

.task, .task-done {
    display: grid;
    grid-template-columns: auto 1fr 1fr auto;
    align-items: center;
    gap: 1rem;
    width: 98%;
    border-top: 1px solid #adaaad;
    padding-top: .7rem;
    padding-bottom: .7rem;
    padding-left: 1rem;
    font-size: .9rem;
}

.task-done {
    text-decoration: line-through;
    text-decoration-thickness: 2px;
    background-color: $primary-grey;
}

.task-content {
    grid-column: 2/2;
    display: flex;
    align-items: center;
    gap: 2rem;
    max-width: 15rem;
    h4 {
        text-overflow: ellipsis;
        overflow-x: hidden;
        white-space: nowrap;
    }
}

.selected-task {
    background-color: $hover-green;
}

.task-properties {
    grid-column: 3/3;
    display: grid;
    grid-template-columns: .2fr .4fr 1fr .2fr;
    grid-template-rows: 1fr;
    align-items: center;

    .end-date, .difficulty, .important, .labels {
        display: flex;
        align-items: center;
        grid-row: 1/1;
    }
    .end-date {
        grid-column: 1/1;
    }

    .difficulty {
        grid-column: 2/2;
    }

    .task-labels {
        grid-column: 3/3;
        font-size: .8rem;
        display: flex;
        flex-wrap: wrap;
        gap: .5rem;
        label {
            border-width: .1rem;
            border-style: solid;
            padding-inline: .2rem;
            border-radius: .5rem;
        }
    }

    .important {
        grid-column: 4/-1;
    }
}

.icon-container {
    grid-column: 1/1;
    display: flex;
    align-items: center;
    cursor: pointer;
    .circle-disappear {
        display: none;
    }
}

.filter-section {
    display: flex;
    justify-content: space-between;
    align-items: center;
    align-content: center;
    flex-direction: row;
    background-color: $primary-grey;
    width: 99%;
    border-radius: .4rem;
    margin-bottom: .5rem;

    .task-counter {
        display: flex;
        align-items: center;
        flex-direction: row;
        gap: 1rem;
        padding-inline: .8rem;
        font-size: .84rem;
    }

    .task-filters {
        padding-right: 2.8rem;
        display: flex;
        align-items: center;
        flex-direction: row;
        gap: 1rem;

        .filter {
            position: relative;
            width: 7rem;
            max-width: 8rem;
            padding-inline: .5rem;
            border-radius: .4rem;
            border: none;
            background-color: transparent;
            display: flex;
        }

        .sort-date, .sort-difficulty {
            font-size: .84rem;
            border: none;
            background-color: transparent;
            &:hover {
                text-decoration: underline;
            }
        }
    }
}