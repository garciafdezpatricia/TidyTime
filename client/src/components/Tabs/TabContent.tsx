import { useTaskContext } from "../Context/TaskContext";
import { MdEdit } from "react-icons/md";
import { GrStar } from "react-icons/gr";
import { IoIosTimer } from "react-icons/io";
import Difficulty from "../DifficultyRate/Difficulty";
import { FaRegCircle, FaCheckCircle } from "react-icons/fa";
import { useEffect, useState } from "react";
import { Label, Task } from "@/src/model/Scheme";
import CheckableComboBox from "../ComboBox/CheckableComboBox";

export interface Props {
    handleCheck: (arg?:any, arg2?:any) => void | any;
    handleEditModal: (arg?:any) => void | any;
    seeDone: boolean,
}

export interface FilterSectionProps {
    handleFilter: (arg?:any) => void | any;
    sortByDate: (arg?:any) => void | any;
    sortByDifficulty: (arg?:any) => void | any;
}

export function FilterSection({handleFilter, sortByDate, sortByDifficulty} : FilterSectionProps) {

    const { tasks, selectedListId } = useTaskContext();
    const [tasksToFilter, setTasksToFilter] = useState<Task[]>([]);
    const [doneTasksCounter, setDoneTasksCounter] = useState(0);
    const [undoneTasksCounter, setUndoneTasksCounter] = useState(0);

    useEffect(() => {
        if (tasks) {
            const allTasks = tasks.map((list) =>{return list.value ? list.value : []}).flat();
            if (allTasks.length > 0) {
                setTasksToFilter(allTasks.filter((task) => task.listIndex === selectedListId))
            } else {
                setTasksToFilter([])
            }
        }
    }, [tasks, selectedListId])

    useEffect(() => {
        let done = 0;
        let undone = 0;
        tasksToFilter.reduce((acc, task) => {
            task.done ? done++ : undone++;
            return null;
        }, null)
        setDoneTasksCounter(done);
        setUndoneTasksCounter(undone);
    }, [tasksToFilter]);

    return (
        <section className="filter-section">
            <div className="task-counter">
                <p>{doneTasksCounter} done</p>
                <p>{undoneTasksCounter} to do</p>
            </div>
            <div className="task-filters">
                <CheckableComboBox 
                    variant="filter"
                    text={"Filter..."} 
                    checkedLabels={[]} 
                    onChange={handleFilter}  
                />
                <button className="sort-date" onClick={sortByDate}>Sort by due date</button>
                <button className="sort-difficulty" onClick={sortByDifficulty}>Sort by difficulty</button>
            </div>
        </section>
    )
}

export default function TabContent({ handleCheck, handleEditModal, seeDone} : Props ) {

    const {tasks, selectedListId, selectedTaskId} = useTaskContext();
    const [tasksToShow, setTasksToShow] = useState(tasks);
    const [filterApplied, setFilterApplied] = useState<Label[]>([]);

    const handleFilter = (filter: Label[]) => {
        if (filter.length > 0 && tasksToShow) {
            const listIndex = tasksToShow?.findIndex((list) => list.key === selectedListId);
            const filteredTaskList = tasksToShow[listIndex].value.filter((task) =>
                task.labels?.some((label) =>
                    filter.some((filterLabel) =>
                        filterLabel.name === label.name && filterLabel.color === label.color
                    )
                )
            ) || [];
            setFilterApplied(filter);
            const mergedTasks = tasks ? [...tasks] : [];
            mergedTasks[listIndex].value = filteredTaskList;
            setTasksToShow(mergedTasks);
        } else {
            setFilterApplied([]);
            setTasksToShow(tasks);
        }
    }

    const sortByDate = () => {
        if (tasks) {
            const listIndex = tasks.findIndex((list) => list.key === selectedListId);
            const orderedTaskList = tasks[listIndex]?.value.
            sort((task1, task2) => {
                if (task1.endDate && task2.endDate) {
                    return new Date(task1.endDate).getTime() - new Date(task2.endDate).getTime();
                } else if (task1.endDate) {
                    return -1; // Las tareas con fecha van primero
                } else if (task2.endDate) {
                    return 1; // Las tareas sin fecha van después
                } else {
                    return 0; // Ambas tareas sin fecha se consideran iguales
                }
            });
        const mergedTasks = [...tasks];
        mergedTasks[listIndex].value = orderedTaskList;
        setTasksToShow(mergedTasks);
        }
    }

    const sortByDifficulty = () => {
        if (tasks) {
            const index = tasks?.findIndex((list) => list.key === selectedListId);
            const orderedTaskList = tasks[index].value.
                sort((task1, task2) => {
                    if (task1.difficulty && task2.difficulty) {
                        return task1.difficulty - task2.difficulty;
                    } else if (task1.difficulty) {
                        return 1; // Las tareas con dificultad van después
                    } else if (task2.difficulty) {
                        return -1; // Las tareas sin dificultad van primero
                    } else {
                        return 0; // Ambas tareas sin dificultad se consideran iguales
                    }
                });
                const mergedTasks = [...tasks];
                mergedTasks[index].value = orderedTaskList;
                setTasksToShow(mergedTasks);
        }
    }

    useEffect(() => {
        if (filterApplied){
            handleFilter(filterApplied);
        } else {
            setTasksToShow(tasks);
        }
    }, [tasks, tasksToShow]);

    return (
        <section className={tasks && tasks.length > 0 ? 'tab-content-container': 'tab-content-empty'}>
            {
                tasks && tasks.length > 0 
                ? <FilterSection handleFilter={handleFilter} sortByDate={sortByDate} sortByDifficulty={sortByDifficulty}/>
                : <p className="empty-tab-title">Add task lists!</p>
            }
				{tasksToShow && tasksToShow.map((content, index) => (
					<ul className={tasksToShow.findIndex((list) => list.key === selectedListId) === index ? "active-content" : "content"} key={index}>
						{content.value && content.value.map((item, itemIndex) => {
							if (item.done && !seeDone) {
								return null;
							}
							return (
								<li
									className={`${item.done ? "task-done" : "task"} ${selectedTaskId === item.id ? "selected-task" : ""}`}
									key={itemIndex}
								>
									<div className='icon-container'>
										<FaRegCircle
											className={item.done ? "circle-disappear" : ""}
											onClick={(e) => handleCheck(item, itemIndex)}
										/>
										<FaCheckCircle
											className={item.done ? "" : "circle-disappear"}
											onClick={(e) => handleCheck(item, itemIndex)}
										/>
									</div>
									<div className='task-content'>
										<h4>{item.title}</h4>
									</div>
									<div className="task-properties">
										<p className="end-date" title={item.endDate} >
											{
												item.endDate && <IoIosTimer size={"1.2rem"} />
											}
										</p>
										<p className="difficulty" title="difficulty">
											<Difficulty difficulty={item.difficulty ?? 0} />
										</p>
										<p className="important" title="important">
											{
												item.important && <GrStar size={"1.2rem"} color="orange"/>
											}
										</p>
										<div className="task-labels">
											{
												item.labels?.map((label, index) => {
													return (
														<label 
															key={index} 
															style={{borderColor: label.color}}>
														{label.name}
														</label>
													)
												})
											}
										</div>
									</div>
									<MdEdit
										className="edit-icon"
										onClick={() => handleEditModal(item.id)}
									/>
								</li>
							);
						})}
					</ul>
				))}
			</section>
    )
}