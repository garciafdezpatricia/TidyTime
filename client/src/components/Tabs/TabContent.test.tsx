import { render } from "@testing-library/react";
import '@testing-library/jest-dom';
import FilterSection from '@/src/components/Tabs/TabContent';
import TabContent from '@/src/components/Tabs/TabContent';

describe('FilterSection', () => {
    it('renders without crashing', () => {
      const { container } = render(<FilterSection handleCheck={() => {}} handleEditModal={() => {}} seeDone={false} />);
      expect(container).toBeInTheDocument();
    });

    it('matches snapshot', () => {
        const { asFragment } = render(<FilterSection handleCheck={() => {}} handleEditModal={() => {}} seeDone={false} />);
        expect(asFragment()).toMatchSnapshot();
      });
});

describe('TabContent', () => {
    it('renders without crashing', () => {
      const { container } = render(<TabContent handleCheck={() => {}} handleEditModal={() => {}} seeDone={false}/>);
      expect(container).toBeInTheDocument();
    });

    it('matches snapshot', () => {
        const { asFragment } = render(<TabContent handleCheck={() => {}} handleEditModal={() => {}} seeDone={false}/>);
        expect(asFragment()).toMatchSnapshot();
      });
});