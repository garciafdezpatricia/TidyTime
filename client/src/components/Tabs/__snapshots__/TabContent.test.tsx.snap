// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterSection matches snapshot 1`] = `
<DocumentFragment>
  <section
    class="tab-content-empty"
    data-testid="tab-content-empty"
  >
    <p
      class="empty-tab-title"
    >
      list.emptyLists
    </p>
  </section>
</DocumentFragment>
`;

exports[`TabContent matches snapshot 1`] = `
<DocumentFragment>
  <section
    class="tab-content-empty"
    data-testid="tab-content-empty"
  >
    <p
      class="empty-tab-title"
    >
      list.emptyLists
    </p>
  </section>
</DocumentFragment>
`;
